cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/CassandraClusterInfo.java

```diff
@@ -338,11 +339,23 @@ public String getVersionFromFeature()
 
     public String getVersionFromSidecar()
     {
-        LOGGER.info("Getting Cassandra versions from all nodes");
-        List<NodeSettings> allNodeSettings = Sidecar.allNodeSettingsBlocking(conf,
-                                                                             cassandraContext.getSidecarClient(),
-                                                                             cassandraContext.clusterConfig);
-        return getLowestVersion(allNodeSettings);
+        List<NodeSettings> currentAllNodeSettings = allNodeSettings;
+        if (currentAllNodeSettings != null)
+        {
+            return getLowestVersion(currentAllNodeSettings);
+        }
+
+        synchronized (this)
+        {
+            if (allNodeSettings == null)
+            {
+                LOGGER.info("Getting Cassandra versions from all nodes");
+                allNodeSettings = Sidecar.allNodeSettingsBlocking(conf,
+                                                                  cassandraContext.getSidecarClient(),
+                                                                  cassandraContext.clusterConfig);
+            }
+            return getLowestVersion(allNodeSettings);
```

Comment:
I think this line could be moved out of the critical section. 
And further more, it can combine with line#345 and making it the classic double checked locking pattern...

Reply:
Refactored.
(I was trying to keep style consistent with other methods within the same class.)

Reply:
To me, the [original implementation provided in the PR](https://github.com/apache/cassandra-analytics/commit/10725fb9f29076637eb2e9a9206d97287b239bc7) is the correct* classic double checked locking pattern. By some definition of correct

Reply:
I agree that getLowestVersion should be out of the critical section